src/Storages/ExternalStream/Kafka/WriteBufferFromKafka.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <atomic>
+#include <librdkafka/rdkafka.h>
+
+#include "IO/BufferWithOwnMemory.h"
+#include "IO/WriteBuffer.h"
+
+namespace DB
+{
+
+using RdKafkaTopicPtr = std::unique_ptr<rd_kafka_topic_t, void (*)(rd_kafka_topic_t *)>;
+
+class WriteBufferFromKafka : public BufferWithOwnMemory<WriteBuffer>
+{
+private:
+    struct State
+    {
+        std::atomic_size_t outstandings;
+        std::atomic_size_t acked;
+        std::atomic_int errorCode;
+    };
+
+public:
+    WriteBufferFromKafka(const Poco::Logger * logger, RdKafkaTopicPtr topic_, size_t buffer_size = DBMS_DEFAULT_BUFFER_SIZE);
+
+    ~WriteBufferFromKafka() override = default;
+
+    void onDrMsg(const rd_kafka_message_t *);
```

Comment:
Let's be a bit more verbose like `onDeliveryMessage` since C++ style is different than Go.

Reply:
Because it's used in `rd_kafka_dr_msg_cb` function call, thus I named it `onDrMsg` to match it (it has nothing to do with Go). But sure, we can make it more verbose.